### ISSUE TITLE
Report upstream the problem of BLAS and LAPACK symbol collision 

### ISSUE THREAD

**Issue by traversaro** [robotology-dependencies/qpOASES#8]
See local fix https://github.com/robotology-dependencies/qpOASES/pull/7 . 
See the similar issue https://github.com/simbody/simbody/issues/510 .

**Comment by traversaro** [comment]
The problem experienced by @GiulioRomualdi was the following:
~~~
#0  dpotrf_ (uplo=<optimized out>, _n=<optimized out>, a=<optimized out>, _lda=<optimized out>, info=0x7fff577ab5f4) at /home/giulio/robot-code/robotology-superbuild/external/qpOASES/src/LAPACKReplacement.cpp:67
#1  0x00007ffff3dc927a in Ipopt::IpLapackDpotrf(int, double*, int, int&) () from /usr/lib/libipopt.so.1
#2  0x00007ffff3dbf442 in Ipopt::DenseGenMatrix::ComputeCholeskyFactor(Ipopt::DenseSymMatrix const&) () from /usr/lib/libipopt.so.1
#3  0x00007ffff3d059a8 in Ipopt::LimMemQuasiNewtonUpdater::UpdateHessian() () from /usr/lib/libipopt.so.1
#4  0x00007ffff3ccaff4 in Ipopt::IpoptAlgorithm::Optimize(bool) () from /usr/lib/libipopt.so.1
#5  0x00007ffff3c4c2c9 in Ipopt::IpoptApplication::call_optimize() () from /usr/lib/libipopt.so.1
#6  0x00007ffff3c4fb40 in Ipopt::IpoptApplication::OptimizeNLP(Ipopt::SmartPtr<Ipopt::NLP> const&, Ipopt::SmartPtr<Ipopt::AlgorithmBuilder>&) () from /usr/lib/libipopt.so.1
#7  0x00007ffff3c42f09 in Ipopt::IpoptApplication::OptimizeNLP(Ipopt::SmartPtr<Ipopt::NLP> const&) () from /usr/lib/libipopt.so.1
#8  0x00007ffff3c4bd2b in Ipopt::IpoptApplication::OptimizeTNLP(Ipopt::SmartPtr<Ipopt::TNLP> const&) () from /usr/lib/libipopt.so.1
#9  0x00007ffff6f8b0c3 in internal::kinematics::InverseKinematicsData::solveProblem() () from /home/giulio/robot-code/robotology-superbuild/build/install/lib/libidyntree-inverse-kinematics.so
#10 0x00000000005edb52 in WalkingIK::computeIK (this=0x90de60, leftTransform=..., rightTransform=..., comPosition=..., result=...) at /home/giulio/robot-code/element_capture-point-walking_right/code/cpp/Walking_module/src/WalkingInverseKinematics.cpp:410
#11 0x00000000005cd57b in WalkingModule::prepareRobot (this=0x7fffffffc0b0) at /home/giulio/robot-code/element_capture-point-walking_right/code/cpp/Walking_module/src/WalkingModule.cpp:1361
#12 0x000000000062d35d in WalkingCommands::read (this=0x7fffffffc178, connection=...) at /home/giulio/robot-code/element_capture-point-walking_right/code/cpp/build/Walking_module/src/WalkingCommands.cpp:147
#13 0x00007ffff50b5e29 in yarp::os::impl::PortCoreAdapter::read(yarp::os::ConnectionReader&) () from /home/giulio/robot-install/lib/libYARP_OS.so.1
#14 0x00007ffff50abaf6 in yarp::os::impl::PortCore::readBlock(yarp::os::ConnectionReader&, void*, yarp::os::OutputStream*) () from /home/giulio/robot-install/lib/libYARP_OS.so.1
#15 0x00007ffff50bb3f8 in yarp::os::impl::PortCoreInputUnit::run() () from /home/giulio/robot-install/lib/libYARP_OS.so.1
#16 0x00007ffff511d8bd in theExecutiveBranch(void*) () from /home/giulio/robot-install/lib/libYARP_OS.so.1
#17 0x00007ffff3739f56 in ACE_OS_Thread_Adapter::invoke() () from /usr/lib/libACE-6.3.3.so
#18 0x00007ffff74116ba in start_thread (arg=0x7fff577ae700) at pthread_create.c:333
#19 0x00007ffff455c41d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
~~~
basically in a binary that linked both qpOASES and BLAS/LAPACK, Ipopt was crashing when calling the qpOASES provided version of `dpotrf_` . Ideally, Ipopt should always call the same version of `dpotrf_` regardess of which libraries are linked in the binary that uses Ipopt.

**Comment by traversaro** [comment]
The problem was fixed upstream with commit https://projects.coin-or.org/qpOASES/changeset/255/ , but it is still unreleased. 

However, I guess the fix was not complete, because the macro `__AVOID_LA_NAMING_CONFLICTS__` is not defined by the CMake-based build system.

**Comment by traversaro** [comment]
Note that the symbols were actually exposed in the public header, due to the following include chain:
`qpOASES.hpp` [-->](https://github.com/robotology-dependencies/qpOASES/blob/ae964551f1bbbfee2e73ca0cfa116f9bc3a388cf/include/qpOASES.hpp#L69) `SQProblemSchur.hpp` [-->](https://github.com/robotology-dependencies/qpOASES/blob/ae964551f1bbbfee2e73ca0cfa116f9bc3a388cf/include/qpOASES/SQProblemSchur.hpp#L44) `LapackBlasReplacement.hpp`  .
The declaration are removed from public headers in changeset https://projects.coin-or.org/qpOASES/changeset/255/ , but I guess this is the rational for not changing the symbol names by default.

**Comment by traversaro** [comment]
@GiulioRomualdi if you ever have a qpOASES library that you need to use that does not incorporate this fix, an easy workaround is just to change the [ELF visibility](https://www.airs.com/blog/archives/42) from `GLOBAL` to `LOCAL` of the conflicting symbols in the qpOASES library directly, using [`objcopy`](https://sourceware.org/binutils/docs/binutils/objcopy.html), with this command:
~~~
objcopy -L sgemm_ -L gemm_ -L spotrf_ -L dpotrf_ -L strtrs_ -L dtrtrs_ -L strcon_ -L dtrcon_ libqpOASES.so 
~~~

**Comment by traversaro** [comment]
@GiulioRomualdi actually the workaround reported in the last comment does not work because you cannot localize symbols in a shared library, see https://bug-binutils.gnu.narkive.com/Dujw1Hw1/objcopy-can-t-localize-dynamic-symbols .

**Comment by traversaro** [comment]
This should be fixed by upstream PR https://github.com/coin-or/qpOASES/pull/108 .

**Comment by traversaro** [comment]
Reported upstream in https://github.com/coin-or/qpOASES/pull/108 .
